[#4060] Align completeness between old Aggregate and new Entity description (#4562)
AxonFramework/AxonFramework: docs/reference-guide/modules/commands/pages/entities/entity-creator.adoc

= Entity Creator
:navtitle: Entity creator

> The `@EntityCreator` annotation tells Axon how to construct the initial entity instance before replaying its events.

Reviewer: Thanks for adding this page, that's create! But, none of this is specific to the `@EntityCreator` annotation. It's behavior that's support both with and without annotations.

Hence, I think it would be good if we first explain the three options we have:

1. No-arg
2. Id-based
3. First-event-based

Added, we should explain that the latter two are a way towards immutable entities (cool new feature of AF5 🎉).

Then, I'd expect sections explaining all three in more detail. Both a declarative and annotation-based sample should be given, of course. Whenever it is annotation-based, you do not have to show the entire entity. Only the bits that are relevant.

It is mandatory for annotation-based entities (autodetected and Spring Boot approaches).

> In the declarative approach, entity construction is handled by the factory passed to `EventSourcedEntityModule.declarative()` instead—see xref:commands:entities/event-sourced-entity.adoc#_declarative[Declarative].

Reviewer 2: the `#_declarative` anchor should be defined explicitly on the heading

[NOTE]
====
The examples below use `@EventSourcedEntity(tagKey = "courseId")Entity`, which is the non-Spring annotation.

> In Spring Boot, replace `@EventSourcedEntity(tagKey = "courseId")Entity` with `@EventSourcedEntity(tagKey = "courseId")`—the `@EntityCreator` patterns are identical.

Reviewer 2: ```suggestion
The examples below use `@EventSourcedEntity(tagKey = "courseId")`, which is the non-Spring annotation.
In Spring Boot, replace `@EventSourcedEntity(tagKey = "courseId")` with `@EventSourcedEntity(tagKey = "courseId")`—the `@EntityCreator` patterns are identical.
```

Each form is supported by both configuration styles:

* **Declarative**: wire the creator with `EventSourcedEntityFactory`.
See xref:commands:entities/event-sourced-entity.adoc#_declarative[Declarative] for the full registration call this slots into.
* **Automated** (annotation-based): mark the constructor with `@EntityCreator`.
Works identically with `@EventSourcedEntity` (non-Spring) and `@EventSourced` (Spring Boot).

[NOTE]
====
The annotated examples below use `@EventSourcedEntity(tagKey = "courseId")`, the non-Spring annotation.
In Spring Boot, replace `@EventSourcedEntity` with `@EventSourced`; the `@EntityCreator` patterns are identical.
====

== No-argument creator

Suited for **mutable** entities.
Axon creates the instance with the no-arg constructor and applies events through `@EventSourcingHandler` methods that mutate the entity's fields.

=== Declarative

[source,java]
----
.entityFactory(c -> EventSourcedEntityFactory.fromNoArgument(CourseEntity::new))
----

> === Automated

Reviewer 2: we are using autodetected in other places for the annotation approach
```suggestion
=== Autodetected
```

Annotate the no-arg constructor with `@EntityCreator`:

[source,java]
----
@EventSourcedEntity(tagKey = "courseId")
public class CourseEntity {

    // state fields, command handlers, event sourcing handlers...

    @EntityCreator // <1>
    protected CourseEntity() {
    }
}
----

<1> Axon invokes this constructor to create the initial instance, then replays all events to rebuild state.

== Identifier-based creator

Suited for entities where the identifier is **always** known at creation time and never changes.
The identifier can be declared `final`, removing a piece of mutable state from the entity.

=== Declarative

[source,java]
----
.entityFactory(c -> EventSourcedEntityFactory.fromIdentifier(courseId -> new CourseEntity(courseId)))
----

> === Automated

Reviewer 2: we are using autodetected in other places for the annotation approach
```suggestion
=== Autodetected
```

<2> Axon invokes this constructor with the entity identifier resolved from the incoming command.

[IMPORTANT]
====
`@InjectEntityId` is required to disambiguate the identifier from an event payload.
Without it, the first parameter is assumed to be the event payload (the first-event-based form below).
====

== First-event-based creator

Suited for **fully immutable** entities: every field is `final`, set once when the entity is created from its creation event.
Subsequent state changes produce a new instance returned from the `@EventSourcingHandler`.

When this form is used, command handlers must be split into xref:commands:entities/event-sourced-entity.adoc#_entity_members[creational and instance command handlers]: a creational handler appends the creation event before the entity exists, and instance handlers operate on the already-created entity for subsequent commands.

=== Declarative

[source,java]
----
.entityFactory(c -> EventSourcedEntityFactory.fromEventMessage((id, firstEvent) -> {
    CourseCreatedEvent created = firstEvent.payloadAs(CourseCreatedEvent.class);
    return new CourseEntity(created.courseId(), created.capacity());
}))
----

> === Automated

Reviewer 2: we are using autodetected in other places for the annotation approach
```suggestion
=== Autodetected
```
